### ISSUE TITLE
False positive not callable for function call in dictionary 

### ISSUE THREAD

**Issue by hbraha** [PyCQA/pylint#5113]
### Bug description

```python
ATTRIBUTES = {
        'DOMAIN': ("domain", str),
        'IMAGE': ("image", str),
}

for key, (name, validate) in ATTRIBUTES.items():
    name = validate(1)
```

### Command used

```shell
pylint test1.py
```

### Pylint output

```shell
************* Module test1
test1.py:8:11: E1102: validate is not callable (not-callable)
```

### Expected behavior

no error.

### Pylint version

```shell
pylint 2.10.0
astroid 2.7.3
Python 3.6.8 (default, Mar 18 2021, 08:58:41)
```

### OS / Environment

Red Hat Enterprise Linux release 8.4

**Comment by mbyrnepr2** [comment]
It seems that when astroid visits the tuple node `key, (name, validate)`, the `newnode` which is produced has `elts` which contain only `name` and `validate` elements (and the key is missing).

Possibly this is due to recursion on this line where the `visit` method recurses into the `visit_tuple` method again and overwrites the `newnode.elts` which at this point contains the `key` object:
https://github.com/PyCQA/astroid/blob/main/astroid/rebuilder.py#L2208

`elts` assigned here: https://github.com/PyCQA/astroid/blob/main/astroid/nodes/node_classes.py#L335

@Pierre-Sassoulas do you think this sounds like the right track?

Aside from this the reported sample code works ok if it is simplified to: 
```
ATTRIBUTES = {
        'DOMAIN': str,
        'IMAGE':  str,
}

for key, validate in ATTRIBUTES.items():
    name = validate(1)
```

In general, pylint doesn't report any issues with the following:
```
for i, j, k, l, m, n, o, p in {"a": "b"}.items():
    print(i, j, k, l, m, n, o, p)
```

**Comment by DanielNoord** [comment]
@mbyrnepr2 That certainly seems like an issue! Good job on identifying it!

**Comment by Pierre-Sassoulas** [comment]
>  the visit method recurses into the visit_tuple method again and overwrites the newnode.elts

This sounds problematic indeed, good job finding this !

**Comment by mbyrnepr2** [comment]
@Pierre-Sassoulas @DanielNoord thanks for looking. 
I take back what I said about the recursion being an issue in my previous comment. On further examination that logic looks ok - I must have confused myself when debugging.

The following changes seem to fix it, but I'll leave as draft for now as I would like to share progress but also double-check when I have more time:
https://github.com/PyCQA/pylint/pull/5593/files
https://github.com/PyCQA/astroid/pull/1311/files

**Comment by mbyrnepr2** [comment]
I think we can close once https://github.com/PyCQA/pylint/pull/5593 is merged (functional test).
